### DISCUSSION TITLE
MagickInverseFourierTransform : How to get magnitue and phase MagickWand representation

### DISCUSSION THREAD

**Discussion by senaki** [ImageMagick/ImageMagick#4253] [Help] [answered]
Hello, 
I'm trying to get the inverse Fourier transform of an image using MagickWand (ImageMagick 6.9.10-23 Q16 x86_64 20190101) function : 

```MagickInverseFourierTransformImage(
  MagickWand *magnitude_wand,MagickWand *phase_wand,
  const MagickBooleanType magnitude)```.
```
The objective is to apply a low-pass filter using C API, equivalent to command-line examples at  https://legacy.imagemagick.org/Usage/fourier/#ft_applications. 

After applying DFT to my source image, I apply  `MagickCompositeImage(fft_image, mask_image, MultiplyCompositeOp, 0,0)`;
Then, I apply the IFT. However,  I don't find how to obtain `magnitude_wand` and `phase_wand` required by `MagickInverseFourierTransformImage`  from my `fft_image` given by `MagickForwardFourierTransformImage`. 

Documentation says for those parameters : 
```
A description of each parameter follows:

magnitude_wand
    the magnitude or real wand. 
phase_wand
    the phase or imaginary wand. 
magnitude
    if true, return as magnitude / phase pair otherwise a real / imaginary image pair. 
```

It tried first to set `magnitude_wand` and `phase_wand` to the same image generated by DFT. 
However, I think the result is not valid. 
I think there is something I don't understand at this point. 

* I know that from command line, FFT generates normally 2 files (magnitude and phase). Is it the same approach when using C API ? if so, how can I do that ?

* I also tried to access the pixel (complex) data to perform the operation, but I don't find how to do that.

Can someone help me or give me a hint to perform such operation ?

Thank you in advance

**Comment by snibgo** [comment, marked as answer]
I don't have time to test this right now, but I think the function you want is:
```
 WandExport MagickBooleanType MagickForwardFourierTransformImage(
  MagickWand *wand,const MagickBooleanType magnitude)
```
... defined in MagickWand/magick-image.c.

If you set magnitude to true, it returns wand->images containing two images: the magnitude, and the phase.

As you say, MagickInverseFourierTransformImage() expects two separate wands, not one wand with two images. So you need to separate the two images into separate wands.

**Reply by senaki** [reply]
Hello, 
thank you for your answer. I am following your suggestion. 
I saw that `wand->images` is updated via  `AddToImageList` by `ForwardFourierTransform`. 1st  magnitude, 2nd phase. 
Since wand->images points to object of type _Image *_,  I tried to convert them into MagickWand object   via : 
```
    MagickWand *magnitude =  NewMagickWandFromImage(GetFirstImageInList(ifft_image->images));
    MagickWand *phase = NewMagickWandFromImage(GetLastImageInList(ifft_image->images));

```

I've got during compilation : 
`error: dereferencing pointer to incomplete type ‘MagickWand’ {aka ‘struct _MagickWand’}`

I only include <wand/MagickWand.h>, is it correct ?

thank you
*EDIT 1*: 
I succeded in dereferencing by including `magick-wand-private.h` but and I correctly get `phase_wand` and `magnitude_wand`. 
However, the IFT still does not gives the correct image.
*EDIT 2*:
It's the mask which is applied that distorts IFT image. Now I obtain, almost a good IFT, except that the mask disk remains. Elsewhere, IFT is correctly blurred.

**Comment by senaki** [comment]
# Solution 
For those who will encounter the same difficulty while beginning with this function. 
Base on the answer https://github.com/ImageMagick/ImageMagick/discussions/4253#discussioncomment-1356546 in this thread. 
Note :<del> You should have included _"wand/magic-wand-private.h"_ in order to properly dereference _MagickWand \*_  objects</del>.
Note : You should use iterators to get first and last image set by MagickForwardFourierTransform.

For example : 
1. Apply ForwardFourierTransForm 
`MagickForwardFourierTransformImage(fft_image, 1); `   
2. Get magnitude and phase MagickWand
```
MagickSetFirstIterator(fft_image);
Im *im = GetImageFromMagickWand(fft_image);
MagickWand *magnitude =  NewMagickWandFromImage(im);
MagickNextImage(fft_image);
im = GetImageFromMagickWand(fft_image);
MagickWand *phase = NewMagickWandFromImage(im);
MagickResetIterator(fft_image);
DestroyImage(im);
```
 3. Generate the mask and apply convolution to magnitude component
```
MagickCompositeImage(magnitude, mask_image, MultiplyCompositeOp, 0,0);

```
4. Apply IFT on convoluted component and resize the image
```
MagickInverseFourierTransformImage(magnitude, phase, 1);
MagickResizeImage(magnitude, w, h, BoxFilter,0 );

```

Hope you'll find it helpful !

**Comment by snibgo** [comment]
> I only include <wand/MagickWand.h>, is it correct ?

Yes.

> I succeded in dereferencing by including magick-wand-private.h ...

I recommend _not_ doing that. Private functions and structures are liable to change. Instead, use functions from magick-wand.c and magick-image.c. For example:

MagickResetIterator() resets the wand iterator.
MagickSetFirstIterator() sets the wand iterator to the first image.
MagickNextImage() sets the next image in the wand as the current image.
GetImageFromMagickWand() returns the current image from the magick wand.
CloneMagickWandFromImages() clones the magick wand and inserts a new image list.

**Reply by senaki** [reply]
I see. I have updated my code and It works using Iterator.
I will update my "solution" answer for other people.

Thanks a lot  @snibgo .
